daemon/stats.go

```diff
@@ -26,26 +32,35 @@ func (daemon *Daemon) ContainerStats(name string, stream bool, out io.Writer) er
 		return ss
 	}
 
-	enc := json.NewEncoder(out)
+	enc := json.NewEncoder(config.OutStream)
 
-	if !stream {
-		// prime the cpu stats so they aren't 0 in the final output
-		s := getStat(<-updates)
+	defer daemon.UnsubscribeToContainerStats(name, updates)
 
-		// now pull stats again with the cpu stats primed
-		s = getStat(<-updates)
-		err := enc.Encode(s)
-		daemon.UnsubscribeToContainerStats(name, updates)
-		return err
-	}
+	noStreamFirstFrame := true
+	for {
+		select {
+		case v, ok := <-updates:
+			if !ok {
+				return nil
+			}
+
+			s := getStat(v)
+			if !config.Stream && noStreamFirstFrame {
+				// prime the cpu stats so they aren't 0 in the final output
+				noStreamFirstFrame = false
+				continue
+			}
+
+			if err := enc.Encode(s); err != nil {
+				// TODO: handle the specific broken pipe
```

Comment:
_Probably_ don't need this comment anymore, since selecting on the channel is what takes most of the time here. If there is a broken pipe, it would most likely be triggered in the select before it ever reached here.